### PR TITLE
ENH: improve t distribution logpdf and pdf for large degrees of freedom

### LINKED ISSUE #18386: BUG: t distribution PDF fails for extreme degrees of freedom
### Describe your issue.

For high `df`, `stats.t.pdf/logpdf` give wrong results.

image

The reason is likely catastrophic cancallation in `sc.gammaln((df+1)/2)-sc.gammaln(df/2)` similar to #18093 .

There are two options to solve this in my opinion: 

- using an asymptotic expansion (see again #18093)
- use Boost's implementation of the t distribution

### Reproducing Code Example

```python
import numpy as np
from scipy.stats import t
import matplotlib.pyplot as plt

df = np.logspace(0, 30, 100)
plt.semilogx(df, t.logpdf(1, df))
plt.xlabel("$df$")
plt.title("$t$ distribution logpdf ($x=1$)")
plt.show()
```

### Error message

```shell
n/a
```

### SciPy/NumPy/Python version and system information

```shell
1.10.1 1.24.2 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
flame_info:
  NOT AVAILABLE
accelerate_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
lapack_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas']
    library_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\lib']
    language = f77
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
blas_info:
    libraries = ['cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\lib']
    include_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\include']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    libraries = ['cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\lib']
    include_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\include']
    language = f77
lapack_opt_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas', 'cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\lib']
    language = f77
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/danie/.conda/envs/scipy1.10\\Library\\include']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
```
```

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Closes #18386 

#### What does this implement/fix?
Implements an asymptotic expansion to work around round off errors. Also, all t distribution tests were moved into the `TestT` class.

#### Additional information

image

<details><summary>Plot generation</summary>
<p>

```python
mp.dps = 500

def asymptotic_formula(x, df):
            return (- 0.5 * (1 + np.log(2 * np.pi)) + df/2 * np.log1p(1/df)
                    + 1/6 * (df + 1)**-1. - 1/45*(df + 1)**-3.
                    - 1/6 * df**-1. + 1/45*df**-3.
                    - (df + 1)/2 * np.log1p(x*x/df))

@np.vectorize
def t_logpdf_mpmath(x, df):
    x = mp.mpf(x)
    df = mp.mpf(df)

    return float(mp.loggamma(0.5 * (df + mp.one)) - mp.loggamma(0.5 * df) - 0.5 * mp.log(df * mp.pi) - 0.5 * (df + mp.one) * mp.log1p(x * x/df))

x = 10.
dfs = np.logspace(1, 5, 1000)
main = stats.t.logpdf(x, dfs)
ref = t_logpdf_mpmath(x, dfs)
pr = asymptotic_formula(x, dfs)

plt.loglog(dfs, np.abs((main - ref)/ref), label="regular", alpha=0.5)
plt.loglog(dfs, np.abs((pr - ref)/ref), label="asymptotic", alpha=0.5)
plt.title("$t$ distribution logpdf: Relative error")
plt.xlabel("$df$")
plt.legend()
``` 

</p>
</details> 

The formula was derived with help of sympy and a bit of further simplifcation.

```python
import sympy
from sympy import log, pi, Rational, simplify, expand

x, a = sympy.symbols('x, a')

def gammaln(x):
    return x*log(x) - x - log(x)/2 + log(2*pi)/2 + 1/(12*x) - 1/(360*x**3)

old = gammaln((1 + x)/2) - gammaln(x/2) - log(pi*x)/2 #+ (x+1)/2 * log(1+(a*a)/x)
print(simplify(old))

# -x*log(x)/2 + x*log(x + 1)/2 - x/(45*(x + 1)**4) - log(pi)/2 - 1/2 - log(2)/2 + 1/(6*x + 6) - 1/(45*(x + 1)**4) - 1/(6*x) + 1/(45*x**3)
``` 

Note that $\ln(x+1)-\ln(x)=\ln(1+1/x)=log1p(1/x)$. Reviewing these expansions can be quite exhausting. The relative error analysis though leaves me at least quite confident that the formula is right.

### CODE REVIEW

**Comment by mdhaber** [comment]
I checked the asymptotic expansion and it looks correct. 
<details>

```python3
# my asymptotic expansion:
import sympy
from sympy import log, pi, Rational, simplify, expand, collect

x = sympy.symbols('x', real=True)
df = sympy.symbols('df', real=True, positive=True)

def gammaln(x):
    return x*log(x) - x - log(x)/2 + log(2*pi)/2 + 1/(12*x) - 1/(360*x**3)

old = gammaln((df+1)/2) - gammaln(df/2) - log(df*pi)/2 - (df+1)/2*log(1+(x**2)/df)
f1 = simplify(simplify(old))

# yours, converted to mpmath
f2 = (- Rational(1, 2) * (1 + log(2 * pi)) + df/2 * log(1+1/df) + Rational(1, 6) * (df + 1)**-1 - Rational(1, 45)*(df + 1)**-3 - Rational(1, 6) * df**-1 + Rational(1, 45)*df**-3 - (df + 1)/2 * log(1+x*x/df))

simplify(f1 - f2)
# 0

```

</details

**Comment by dschmitz89** [comment]
> I checked the asymptotic expansion and it looks correct.
> 
> You seemed to agree with the `ReferenceDistribution` idea. Why not use it here so that if we ever need reference value for the t distribution, we'll have them?

The reference distribution worked quite easily. Thanks for the hint, I had already forgotten about it, habits die slowly ;).

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -3185,98 +3185,110 @@ def test_stats_broadcast(self):
...
+
+    @pytest.mark.parametrize("x, df, logpdf_ref, pdf_ref",
+                             [(1, 1e100,
+                               -1.4189385332046727, 0.24197072451914334),
+                              (1e3, 1e50, -500000.9189385332, 0.),
+                              (10, 1e20,
+                               -50.918938533204674, 7.69459862670642e-23),
+                              (1, 1,
+                               -1.8378770664093456, 0.15915494309189535)])
+    def test_logpdf_pdf(self, x, df, logpdf_ref, pdf_ref):
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:3289]: Optional: In the past, the style has been to parameterize these tests as you have done. Parameterization is definitely preferable to adding explicit loops in tests, and it is nice if you want it to be very explicit which part of a test is failing. I'm not sure if we need either of these features in these sorts of tests, though. What do you think about defining 

```suggestion
    # caution: I haven't checked this locally for execution or PEP8
    def test_logpdf_pdf(self, x, df, logpdf_ref, pdf_ref):
        x = [1, 1e3, 10, 1]
        df = [1e100, 1e50, 1e20, 1]
        logpdf_ref = [-1.4189385332046727, -500000.9189385332,
                      -50.918938533204674, -1.8378770664093456]
        pdf_ref = [0.24197072451914334, 0, 7.69459862670642e-23, 
                   0.15915494309189535]
```
It's a little more compact and probably slightly faster to run, but the main advantage in my opinion is easy copy-paste into a script to test locally.

What do you think?

**Review by mdhaber — APPROVED** [review]
This looks good to me; I just want to run the reference values locally.

File: scipy/stats/tests/test_distributions.py (hunk 1 above)
Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:3289]: I always like the output of `pytest -v` with parametrized tests but that's my personal preference.

What I see typically in new tests of these types of functionality in scipy are usually parametrized tests. Not sure if there are guidelines about it in the docs but we should have some level of uniformity in my opinion. So I would prefer to leave the tests as they are here.

**Pull request merged**